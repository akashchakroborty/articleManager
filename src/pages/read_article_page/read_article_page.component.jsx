import React, { Fragment, useEffect, useState } from "react";
import {
  ReadArticlePageContainer,
  Back,
  ReadMorePreviewContainer,
  ReadMoreLabel,
} from "./read_article_page.styles";
import ArticleRead from "../../components/article-read/article-read.component";
import ArticleReadMore from "../../components/article_read_more/article_read_more.component";
import NotFound from "../../components/notfound/notfound.component";
import WithSpinner from "../../components/with-spinner/with-spinner.component";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import {
  selectArticles,
  selectIsFetching,
  selectArticle,
} from "../../redux/articles/articles.selectors";

export const ReadArticlePage = ({
  articles,
  article,
  match,
  history,
  isLoading,
}) => {
  const validateArticleReadRoute = () => {
    const validIds = articles.map((val) => val.id);
    return validIds.includes(match.params.id);
  };
  const [shuffledArticles, setShuffledArticles] = useState([]);

  useEffect(() => {
    setShuffledArticles(
      [...articles].sort(() => 0.5 - Math.random()).slice(0, 3)
    );
  }, []);

  return validateArticleReadRoute() ? (
    <Fragment>
      <ReadArticlePageContainer>
        <ArticleRead article={article} />
        <Back className="back-button" onClick={() => history.push("/")}>
          BACK
        </Back>
      </ReadArticlePageContainer>
      <ReadMoreLabel className="read-more">More from ReadIt</ReadMoreLabel>
      <ReadMorePreviewContainer className="read-more-container">
        {shuffledArticles.map((article) => (
          <ArticleReadMore key={article.id} article={article} />
        ))}
      </ReadMorePreviewContainer>
    </Fragment>
  ) : (
    <NotFound history={history} />
  );
};

const mapStateToProps = createStructuredSelector({
  articles: selectArticles,
  isLoading: selectIsFetching,
  article: selectArticle,
});

export default connect(mapStateToProps)(WithSpinner(ReadArticlePage));
