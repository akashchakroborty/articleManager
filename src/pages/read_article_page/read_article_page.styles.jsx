import styled from "styled-components";

export const ReadArticlePageContainer = styled.div`
  width: 850px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin: 30px auto;
  @media screen and (max-width: 800px) {
    flex-direction: column;
    width: unset;
    align-items: center;
    > *:first-child {
      margin-bottom: 50px;
    }
  }
`;

export const Back = styled.div`
  text-decoration: none;
  border: 1px solid blue;
  cursor: pointer;
  border-radius: 4px;
  padding: 4px 12px;
  width: fit-content;
  align-self: center;
  color: blue;
  letter-spacing: 3px;
  font-weight: bold;
  line-height: 30px;
`;

Back.displayName = "Back";

export const ReadMorePreviewContainer = styled.div`
  display: flex;
  margin: 20px;
  justify-content: space-evenly;
  @media screen and (max-width: 800px) {
    flex-wrap: wrap;
  }
`;
export const ReadMoreLabel = styled.div`
  text-align: left;
  margin: 20px 4vw;
  padding-bottom: 8px;
  margin-bottom: 32px;
  font-weight: bold;
  color: gray;
  font-size: 28px;
  border-bottom: 1px solid rgba(230, 230, 230, 1);
`;
