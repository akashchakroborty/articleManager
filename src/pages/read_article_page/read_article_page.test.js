import React from "react";
import { shallow } from "enzyme";
import { ReadArticlePage } from "./read_article_page.component";

describe("ReadArticlePage component", () => {
  let wrapper;
  let mockHistory;
  let mockMatch;
  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };
    mockMatch = {
      params: {
        id: "1234",
      },
    };
    const mockProps = {
      articles: [
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "1234",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "12",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "1345",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
      ],
      article: {
        title: "Mock data",
        description: "Mock data",
        author: "Mock data",
        content: "Mock data",
        id: "1234",
        img: "Mock data",
        avatar: "Mock data",
        claps: "Mock data",
      },
      history: mockHistory,
      match: mockMatch,
    };

    wrapper = shallow(<ReadArticlePage {...mockProps} />);
  });

  it("should render ReadArticlePage component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call history.push when Back is clicked", () => {
    wrapper.find("Back").simulate("click");
    expect(mockHistory.push).toHaveBeenCalled();
  });

  it("should render Not Found when we have an invalid route.", () => {
    const newMockProps = {
      articles: [
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "12",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "123",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "1267",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
      ],
      article: {
        title: "Mock data",
        description: "Mock data",
        author: "Mock data",
        content: "Mock data",
        id: "12",
        img: "Mock data",
        avatar: "Mock data",
        claps: "Mock data",
      },
      history: mockHistory,
      match: mockMatch,
    };
    const newWrapper = shallow(<ReadArticlePage {...newMockProps} />);

    expect(newWrapper.exists("NotFound")).toBe(true);
  });
});
