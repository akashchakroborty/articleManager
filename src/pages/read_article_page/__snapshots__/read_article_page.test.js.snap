// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadArticlePage component should render ReadArticlePage component 1`] = `ShallowWrapper {}`;
